### PR TITLE
Add Django ASGI support

### LINKED ISSUE #165: Django ASGI not supported
**Describe your environment**
Python 3.8
Docker(python:3.8 image)
daphne-2.5.0/uvicorn 0.11.8/hypercorn 0.10.2
Django-3.1.0

**Steps to reproduce**
```sh
git clone https://github.com/HiveTraum/opentelemetry-django-asgi-example
cd opentelemetry-django-asgi-example
docker-compose up -d
curl -I 127.0.0.1:8000/api/
curl -I 127.0.0.1:8001/api/
curl -I 127.0.0.1:8002/api/
```

**What is the expected behavior?**
Correct request processing

**What is the actual behavior?**
KeyError when trying to get `wsgi.url_scheme` from environ

**Additional context**

Every wsgi server fill this variable inside request.META, but every asgi doesn't. Actually I don't know precisely to whom this bug can be  addressed. For now opentelemetry-instrumentation-django relies on opentelemetry-instrumentation-wsgi, but ignores opentelemetry-instrumentation-asgi, maybe this can be key to solve this issue

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/asgiref/sync.py", line 330, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 38, in inner
    response = await get_response(request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 221, in _get_response_async
    response = await middleware_method(request, callback, callback_args, callback_kwargs)
  File "/usr/local/lib/python3.8/site-packages/asgiref/sync.py", line 296, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/usr/local/lib/python3.8/asyncio/tasks.py", line 455, in wait_for
    return await fut
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/site-packages/asgiref/sync.py", line 334, in thread_handler
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/instrumentation/django/middleware.py", line 73, in process_view
    attributes = collect_request_attributes(environ)
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/instrumentation/wsgi/__init__.py", line 111, in collect_request_attributes
    result["http.url"] = wsgiref_util.request_uri(environ)
  File "/usr/local/lib/python3.8/wsgiref/util.py", line 72, in request_uri
    url = application_uri(environ)
  File "/usr/local/lib/python3.8/wsgiref/util.py", line 52, in application_uri
    url = environ['wsgi.url_scheme']+'://'
KeyError: 'wsgi.url_scheme'
```

### PR DESCRIPTION
**Pull request by adamantike**
# Description

Still a WIP, this diff adds `asgi` as an extra, and uses its methods if the current request is an `ASGIRequest`.

I still need to dig deeper in the current test suite, to find a way to duplicate the tests in
`instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py`, but using an [`AsyncClient`](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#testing-asynchronous-code).

Fixes #165, #185, #280, #334.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by adamantike** [comment]
Marking this as ready for review. There are still a few things to investigate:

* Tests are passing, but triggering log error: `opentelemetry.context:__init__.py:146 Failed to detach context`.
* ASGI intrumentation's `set_status_code` method does not populate the `status_text` field. I think this should be tackled in a separate PR.
* Traced request attributes are not supported in this PR yet.

cc @lonewolf3739, because of the context provided in #334.

**Comment by ocelotl** [comment]
@adamantike is this still WIP? If so, please mark it as draft.

**Comment by ocelotl** [comment]
> @adamantike is this still WIP? If so, please mark it as draft.

Oh, I see, please edit the PR description instead.

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+    async def test_span_name_404(self):
+        await self.async_client.get("/span_name/1234567890/")
+        span_list = self.memory_exporter.get_finished_spans()
+        self.assertEqual(len(span_list), 1)
+
+        span = span_list[0]
+        self.assertEqual(span.name, "HTTP GET")
+
+    @pytest.mark.skip(
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py:265]: I think it would be better not to have this test at all instead of skipping it every time and add it after traced request attributes are supported. That way it won't show up constantly as being skipped in the test runs.

**Review by ocelotl — CHANGES_REQUESTED** [review]

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py [hunk 2]
```diff
@@ -113,7 +113,7 @@ def get_host_port_url_tuple(scope):
     """
     server = scope.get("server") or ["0.0.0.0", 80]
     port = server[1]
-    server_host = server[0] + (":" + str(port) if port != 80 else "")
+    server_host = server[0] + (":" + str(port) if str(port) != "80" else "")
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:116]: I found this discrepancy in [Django's `AsyncRequestFactory`](https://github.com/django/django/blob/a948d9df394aafded78d72b1daa785a0abfeab48/django/test/client.py#L500) while testing, as the base scope is created with a port `'80'`, instead of using an integer.

I think that changing this check could avoid any issues with other potential callers using a string instead of an integer.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 3]
```diff
@@ -182,7 +230,10 @@ def process_response(self, request, response):
             request.META.pop(self._environ_activation_key)
 
         if self._environ_token in request.META.keys():
-            detach(request.environ.get(self._environ_token))
+            if is_asgi_request:
+                detach(request.META.get(self._environ_token))
+            else:
+                detach(request.environ.get(self._environ_token))
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:236]: Consider that `request.environ` is only [defined in `WSGIRequest`](https://github.com/django/django/blob/a948d9df394aafded78d72b1daa785a0abfeab48/django/core/handlers/wsgi.py#L70), but not set in `ASGIRequest`, so moving this line out of the `else` block would trigger an `AttributeError` if we're processing an asynchronous request and try to execute `detach(request.environ.get(self._environ_token))`.

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+    @classmethod
+    def _add_databases_failures(cls):
+        # Disable databases.
+        pass
+
+    @classmethod
+    def _remove_databases_failures(cls):
+        # Disable databases.
+        pass
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py:112]: The purpose of these methods was to override Django's provided ones, which try to validate and database connections: https://github.com/django/django/blob/a948d9df394aafded78d72b1daa785a0abfeab48/django/test/testcases.py#L202-L226

I'll recheck if they are still needed, or if we can delete them, though.

[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py]: Found the methods unnecessary after avoiding inheriting from `SimpleTestCase` class from Django, and correctly setting up Django settings in `setUpClass`!

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py (hunk 1 above)
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py]: I have added support for traced request attributes, so unskipped the test instead :)

**Comment by adamantike** [comment]
I have rebased, fixed some comments, and added support for traced request attributes! The only issue I still find is this one:

* Tests are passing, but triggering log error: opentelemetry.context:__init__.py:146 Failed to detach context.

However, it could be related to how Pytest handles ContextVars and contexts related to attached tokens, based on https://github.com/pytest-dev/pytest-asyncio/issues/127

**Review by ocelotl — APPROVED** [review]

**Comment by adamantike** [comment]
Please do not merge yet. I just found that the `Failed to detach context` log messages also appear ina real Django project, when manually installing this branch, for a project using Gunicorn with Uvicorn workers.

**Comment by adamantike** [comment]
I have filed https://code.djangoproject.com/ticket/32815, to determine if the bug is in Django's codebase instead.

**Comment by adamantike** [comment]
Filed https://github.com/open-telemetry/opentelemetry-python-contrib/pull/533, which circumvents the `Failed to detach context` error by simplifying how the Django middleware is wrapped with `sync_to_async`. With that change applied, the issue doesn't occur anymore!

**Comment by adamantike** [comment]
Rebased now that #533 was merged. All the `Failed to detach context` errors have been fixed, and this is now ready for another review!

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 5]
```diff
@@ -69,6 +70,26 @@ def __call__(self, request):
         MiddlewareMixin = object
 
 
+try:
+    from django.core.handlers.asgi import ASGIRequest
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:74]: Better to check for version instead, an import error may not be caused necessarily by this component being missing.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 6]
```diff
@@ -92,6 +113,10 @@ def __call__(self, request):
 ]
 
 
+def _is_asgi_request(request: HttpRequest) -> bool:
+    return bool(ASGIRequest and isinstance(request, ASGIRequest))
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:117]:
```suggestion
    return ASGIRequest is not None and isinstance(request, ASGIRequest)
```

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 7]
```diff
@@ -156,12 +192,36 @@ def process_request(self, request):
...
+                # ASGI requests include extra attributes in request.scope.headers. For this reason,
+                # we need to build an object with the union of `request` and `request.scope.headers`
+                # contents, for the extract_attributes_from_object function to be able to retrieve
+                # attributes from it.
+                attributes = extract_attributes_from_object(
+                    types.SimpleNamespace(
+                        **{
+                            **request.__dict__,
+                            **{
+                                name.decode("latin1"): value.decode("latin1")
```
Comment (ocelotl) [inline instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:211]: Just curious, why `"latin1"`?

**Review by ocelotl — CHANGES_REQUESTED** [review]
Looking good! Just minor changes requested.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py (hunk 5 above)
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:74]: I agree, also maintains consistency with the version check we are now doing to decide the Middleware style being used.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py (hunk 7 above)
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:211]: Here, I'm using the same logic Django uses to decode the request scope in ASGI mode, which uses `latin1`. This is the relevant upstream code: https://github.com/django/django/blob/7bb0605a8c4adb4247a179276036f138ae859ac9/django/core/handlers/asgi.py#L76-L90

**Comment by adamantike** [comment]
@ocelotl, thanks for the feedback! Pushed the requested changes.

**Review by ocelotl — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-django/setup.cfg [hunk 8]
```diff
@@ -45,6 +45,8 @@ install_requires =
     opentelemetry-semantic-conventions == 0.23.dev0
 
 [options.extras_require]
+asgi =
+    opentelemetry-instrumentation-asgi == 0.23.dev0
```
Comment (owais) [inline instrumentation/opentelemetry-instrumentation-django/setup.cfg:49]: This should be documented so people using ASGI with Django realize they need to install these extra deps to make it work. 

How do ASGI deps work with Django? Do regular Django projects need to install additional packages to make ASGI work or does Django now ship with (or depend on) bundled ASGI module.  If some level of ASGI support is always present in Django now, may be we can add this as a main dependency instead?

**Review by owais — CHANGES_REQUESTED** [review]

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 9]
```diff
@@ -92,6 +113,10 @@ def __call__(self, request):
 ]
 
 
+def _is_asgi_request(request: HttpRequest) -> bool:
+    return ASGIRequest is not None and isinstance(request, ASGIRequest)
```
[the remark this answers is not in the document]
Comment (ocelotl) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:117]: `ASGIRequest` can be `None`, in that case this suggested approach will cause an error:

```python
In [1]: isinstance(1, None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-ce5efb9b0eda> in <module>
----> 1 isinstance(1, None)

TypeError: isinstance() arg 2 must be a type or tuple of types
```

Also, if checking that a certain object is not `None` the `is not None` form [should be used](https://www.python.org/dev/peps/pep-0008/#programming-recommendations).

[the remark this answers is not in the document]
Comment (owais) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:117]: I think there is also an is_asgi_supported variable being set above. we could also use that and then run instance check

**Comment by owais** [comment]
@adamantike Hi, thanks for taking this up. Are you still planning to finish this?

**Comment by adamantike** [comment]
@owais, I'll be tackling your comments this week, so we can move forward! Thanks for all the feedback

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:116]: I don't think we should mix the concepts of `is_asgi_supported` (based on required Python packages being installed) from `_is_asgi_request` (based on the current request being handled), considering that the Django middleware can handle both WSGI and ASGI requests, and most of the conditionals are `if is_asgi_request and not _is_asgi_supported` to return early for invalid tracing scenarios.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 10]
```diff
@@ -141,12 +166,23 @@ def process_request(self, request):
         if self._excluded_urls.url_disabled(request.build_absolute_uri("?")):
             return
 
+        is_asgi_request = _is_asgi_request(request)
+        if is_asgi_request and not _is_asgi_supported:
+            return
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:170]: > Can a django service/middleware handle both WSGI and ASGI requests at the same time?
Yes, it can handle both, so we need to check which kind of request we're receiving on each call.

> At least is_asgi_supported can be called before _is_asgi_request and that way we can avoid running the type check for non-ASGI requests.
As we're using `is_asgi_request` multiple times on this method, the code will need to be a bit more convoluted if we want to avoid the type check when we know ASGI is not supported. I will change the code accordingly, and we can decide if it's worth to keep it that way.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 11]
```diff
@@ -156,12 +192,36 @@ def process_request(self, request):
             ),
         )
 
-        attributes = collect_request_attributes(request_meta)
+        if is_asgi_request:
+            attributes = collect_request_attributes(request.scope)
+        else:
+            attributes = collect_request_attributes(request_meta)
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:197]: Good idea, applied!

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 12]
```diff
@@ -156,12 +192,36 @@ def process_request(self, request):
...
 
         if span.is_recording():
-            attributes = extract_attributes_from_object(
-                request, self._traced_request_attrs, attributes
-            )
+            if is_asgi_request:
+                # ASGI requests include extra attributes in request.scope.headers. For this reason,
+                # we need to build an object with the union of `request` and `request.scope.headers`
+                # contents, for the extract_attributes_from_object function to be able to retrieve
+                # attributes from it.
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:204]: Would it make more sense, to keep using the `extract_attributes_from_object` util, to go with a composition approach?

```python
attributes = extract_attributes_from_object(
    request, self._traced_request_attrs, attributes
)
if is_asgi_request:
    # ASGI requests include extra attributes in request.scope.headers.
    attributes = extract_attributes_from_object(
        types.SimpleNamespace(**{
            name.decode("latin1"): value.decode("latin1")
            for name, value in request.scope.get("headers", [])
        }),
        self._traced_request_attrs,
        attributes,
    )
```

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 13]
```diff
@@ -208,15 +268,24 @@ def process_response(self, request, response):
...
         activation = request.META.pop(self._environ_activation_key, None)
         span = request.META.pop(self._environ_span_key, None)
 
         if activation and span:
-            add_response_attributes(
-                span,
-                "{} {}".format(response.status_code, response.reason_phrase),
-                response,
-            )
+            if is_asgi_request:
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:278]: As the functions are right now, I don't see any needed changes for both WSGI and ASGI to make the same changes to the current span:
* WSGI: https://github.com/open-telemetry/opentelemetry-python-contrib/blob/e12917401f5ca6af8737378a99b00da2b1987d5c/instrumentation/opentelemetry-instrumentation-wsgi/src/opentelemetry/instrumentation/wsgi/__init__.py#L188-L208
* ASGI: https://github.com/open-telemetry/opentelemetry-python-contrib/blob/e12917401f5ca6af8737378a99b00da2b1987d5c/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L125-L140

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 14]
```diff
@@ -208,15 +268,24 @@ def process_response(self, request, response):
...
+                set_status_code(span, response.status_code)
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:279]: I also think this is already fixed in the current `set_status_code` implementation.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 15]
```diff
@@ -68,6 +70,25 @@ def __call__(self, request):
     except ImportError:
         MiddlewareMixin = object
 
+if DJANGO_3_0:
+    from django.core.handlers.asgi import ASGIRequest
+else:
+    ASGIRequest = None
+
+try:
```
[the remark this answers is not in the document]
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:78]: Applied, thanks!

File: instrumentation/opentelemetry-instrumentation-django/setup.cfg (hunk 8 above)
Comment (adamantike) [reply instrumentation/opentelemetry-instrumentation-django/setup.cfg]: This is a good discussion topic. Django doesn't require any additional/extra dependencies for ASGI to work (aside from running it using Uvicorn/Daphne), and bundles `asgiref` since the async introduction in Django 3.0: https://github.com/django/django/commit/a415ce70bef6d91036b00dd2c8544aed7aeeaaed#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7

This dependency doesn't include other heavy dependencies that aren't already required by `opentelemetry-instrumentation-wsgi`. However, we also have to consider that ASGI is only present since Django 3.x, while this project tries to maintain support for older Django versions.

I think the only missing change on this PR, if we go with adding this to `install_requires`, is to make `_is_asgi_supported` also depend on `DJANGO_3_0` being `True`, as we won't receive any `ImportError` anymore when importing ASGI dependencies.

**Comment by adamantike** [comment]
@owais @ocelotl, sorry for the delay. This is now ready for another review!

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:305]: How come we can't `detach(request.META.get(self._environ_token))` for both?

**Review by lzchen — APPROVED** [review]

**Comment by owais** [comment]
@adamantike sorry for the delay. can you please resolve the conflicts and rebase/merge with main? We can try to include it in the release this week.

**Comment by adamantike** [comment]
I agree, though by that same approach, a future breaking change should be introduced for `wsgi` to also become an extra, so users exclusively having ASGI don't get WSGI dependencies.

This is where https://github.com/pypa/setuptools/issues/1139 would come useful if implemented. If no extras are specified, you get both WSGI and ASGI. If you want to cut down on dependencies, you need to specify which flow your app is going to use.

**Comment by owais** [comment]
Oh I've been waiting on pypa/setuptools#1139 for longer than I remember. It's such an obvious feature to have :)
